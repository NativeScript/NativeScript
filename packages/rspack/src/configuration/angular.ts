import { existsSync } from 'fs';
import { extname, resolve } from 'path';
import Config from 'rspack-chain';
import { satisfies } from 'semver';

import { getDependencyVersion } from '../helpers/dependencies';
import { warnOnce } from '../helpers/log';
import {
	getEntryDirPath,
	getEntryPath,
	getPlatformName,
} from '../helpers/platform';
import { getProjectTSConfigPath } from '../helpers/project';
import { getTypescript, readTsConfig } from '../helpers/typescript';
import { env as _env, IWebpackEnv } from '../index';
import base from './base';

// until we can switch to async/await on the webpack config, copy this from '@angular/compiler-cli'
const GLOBAL_DEFS_FOR_TERSER = {
	ngDevMode: false,
	ngI18nClosureMode: false,
};

const GLOBAL_DEFS_FOR_TERSER_WITH_AOT = {
	...GLOBAL_DEFS_FOR_TERSER,
	ngJitMode: false,
};

export default function (config: Config, env: IWebpackEnv = _env): Config {
	base(config, env);

	const platform = getPlatformName();
	const tsConfigPath = getProjectTSConfigPath();
	const disableAOT = !!env.disableAOT;

	// remove default ts rule
	config.module.rules.delete('ts');

	// remove fork ts checked as not needed
	config.plugins.delete('ForkTsCheckerWebpackPlugin');

	// explicitly define mainFields to make sure ngcc compiles as es2015 (module field)
	// instead of umd (main field).
	config.resolve.mainFields.add('module').add('main');

	config.module
		.rule('angular')
		.test(/(?:\.ngfactory.js|\.ngstyle\.js|\.ts)$/)
		.use('@ngtools/webpack')
		.loader('@ngtools/webpack');

	config.module
		.rule('@angular/core')
		.test(/[\/\\]@angular[\/\\]core[\/\\].+\.js$/)
		.parser({ system: true });

	// set up html
	config.module
		.rule('html')
		.test(/\.html$/)
		.use('raw-loader')
		.loader('raw-loader');

	// exclude component css files from the normal css rule
	config.module
		.rule('css')
		.include.add(resolve(getEntryDirPath(), 'app.css'))
		.add(resolve(getEntryDirPath(), `app.${platform}.css`))
		.add(/node_modules/);

	// and instead use raw-loader, since that's what angular expects
	config.module
		.rule('css|component')
		.exclude.add(resolve(getEntryDirPath(), 'app.css'))
		.add(resolve(getEntryDirPath(), `app.${platform}.css`))
		.add(/node_modules/)
		.end()
		.test(/\.css$/)
		.use('raw-loader')
		.loader('raw-loader');

	// get base postCSS options
	const postCSSOptions = config.module
		.rule('scss')
		.uses.get('postcss-loader')
		.get('options');

	// exclude component css files from the normal css rule
	config.module
		.rule('scss')
		.include.add(resolve(getEntryDirPath(), 'app.scss'))
		.add(resolve(getEntryDirPath(), `app.${platform}.scss`))
		.add(/node_modules/);

	// and instead use raw-loader, since that's what angular expects
	config.module
		.rule('scss|component')
		.exclude.add(resolve(getEntryDirPath(), 'app.scss'))
		.add(resolve(getEntryDirPath(), `app.${platform}.scss`))
		.add(/node_modules/)
		.end()
		.test(/\.scss$/)
		.use('raw-loader')
		.loader('raw-loader')
		.end()
		.use('postcss-loader')
		.loader('postcss-loader')
		.options(postCSSOptions)
		.end()
		.use('sass-loader')
		.loader('sass-loader');

	const angularCompilerPlugin = getAngularCompilerPlugin();
	if (angularCompilerPlugin) {
		config.plugin('AngularCompilerPlugin').use(angularCompilerPlugin, [
			{
				tsConfigPath,
				mainPath: getEntryPath(),
				// disable type checking in a forked process - it ignores
				// the hostReplacementPaths and prints errors about
				// platform suffixed files, even though they are
				// working as expected.
				forkTypeChecker: false,
				hostReplacementPaths(path: string) {
					const ext = extname(path);
					const platformExt = `.${platform}${ext}`;

					// already includes a platform specific extension - ignore
					if (path.includes(platformExt)) {
						return path;
					}

					const platformPath = path.replace(ext, platformExt);
					// check if the same file exists with a platform suffix and return if it does.
					if (existsSync(platformPath)) {
						// console.log(`[hostReplacementPaths] resolving "${path}" to "${platformPath}"`);
						return platformPath;
					}

					// just return the original path otherwise
					return path;
				},
				platformTransformers: [require('../transformers/NativeClass').default],
			},
		]);
	}

	const angularWebpackPlugin = getAngularWebpackPlugin();
	if (angularWebpackPlugin) {
		// angular no longer supports transformers.
		// so we patch their method until they do
		// https://github.com/angular/angular-cli/pull/21046
		const originalCreateFileEmitter =
			angularWebpackPlugin.prototype.createFileEmitter;
		angularWebpackPlugin.prototype.createFileEmitter = function (
			...args: any[]
		) {
			let transformers = args[1] || {};
			if (!transformers.before) {
				transformers.before = [];
			}
			if (this.pluginOptions.jitMode) {
				transformers.before.unshift(
					require('../transformers/NativeClass').default,
				);
			} else {
				transformers.before.push(
					require('../transformers/NativeClass').default,
				);
			}
			args[1] = transformers;
			return originalCreateFileEmitter.apply(this, args);
		};
		config.plugin('AngularWebpackPlugin').use(angularWebpackPlugin, [
			{
				tsconfig: tsConfigPath,
				directTemplateLoading: false,
				jitMode: disableAOT,
			},
		]);

		config.when(env.hmr, (config) => {
			config.module
				.rule('angular-hmr')
				.enforce('post')
				.test(getEntryPath())
				.use('angular-hot-loader')
				.loader('angular-hot-loader');
		});

		const buildAngularVersion = getDependencyVersion(
			'@angular-devkit/build-angular',
		);

		if (buildAngularVersion) {
			const buildAngularOptions: any = {
				aot: !disableAOT,
			};

			if (satisfies(buildAngularVersion, '<15.0.0')) {
				const tsConfig = readTsConfig(tsConfigPath);
				const { ScriptTarget } = getTypescript();
				buildAngularOptions.scriptTarget =
					tsConfig.options.target ?? ScriptTarget.ESNext;
			}

			if (disableAOT) {
				buildAngularOptions.optimize = false;
			}
			// zone + async/await
			config.module
				.rule('angular-webpack-loader')
				.test(/\.[cm]?[tj]sx?$/)
				.exclude.add(
					/[/\\](?:core-js|@babel|tslib|web-animations-js|web-streams-polyfill)[/\\]/,
				)
				.end()
				.resolve.set('fullySpecified', false)
				.end()
				.before('angular')
				.use('webpack-loader')
				.loader(getWebpackLoaderPath())
				.options(buildAngularOptions);
		} else {
			warnOnce(
				'build-angular-missing',
				`
				@angular-devkit/build-angular is missing! Some features may not work as expected. Please install it manually to get rid of this warning.
				`,
			);
		}
	}

	// look for platform specific polyfills first
	// falling back to independent polyfills
	const polyfillsPath = [
		resolve(getEntryDirPath(), `polyfills.${platform}.ts`),
		resolve(getEntryDirPath(), `polyfills.ts`),
	].find((path) => existsSync(path));

	if (polyfillsPath) {
		const paths = config.entry('bundle').values();

		// replace globals with the polyfills file which
		// should handle loading the correct globals
		// and any additional polyfills required.
		if (paths.includes('@nativescript/core/globals/index')) {
			paths[paths.indexOf('@nativescript/core/globals/index')] = polyfillsPath;

			// replace paths with the updated paths
			config.entry('bundle').clear().merge(paths);
		}
	}

	// Filter common undesirable warnings
	config.set(
		'ignoreWarnings',
		(config.get('ignoreWarnings') ?? []).concat([
			/**
			 * This rule hides
			 * +-----------------------------------------------------------------------------------------+
			 * | WARNING in Zone.js does not support native async/await in ES2017+.                      |
			 * | These blocks are not intercepted by zone.js and will not triggering change detection.   |
			 * | See: https://github.com/angular/zone.js/pull/1140 for more information.                 |
			 * +-----------------------------------------------------------------------------------------+
			 */
			/Zone\.js does not support native async\/await/,
			/**
			 * This rule hides
			 * +-----------------------------------------------------------------------------------------+
			 * | WARNING in environment.*.ts is part of the TypeScript compilation but it's unused.      |
			 * | Add only entry points to the 'files' or 'include' properties in your tsconfig.          |
			 * +-----------------------------------------------------------------------------------------+
			 */
			/environment(\.(\w+))?\.ts is part of the TypeScript compilation but it's unused/,
			// Additional rules to suppress warnings that are safe to ignore
			{
				module: /@angular\/core\/(__ivy_ngcc__\/)?fesm2015\/core.js/,
				message:
					/Critical dependency: the request of a dependency is an expression/,
			},
			/core\/profiling/,
			/core\/ui\/styling/,
		]),
	);

	config.optimization.minimizer('TerserPlugin').tap((args) => {
		args[0].terserOptions ??= {};
		args[0].terserOptions.compress ??= {};
		args[0].terserOptions.compress.global_defs ??= {};
		args[0].terserOptions.compress.global_defs = {
			...args[0].terserOptions.compress.global_defs,
			...(disableAOT
				? GLOBAL_DEFS_FOR_TERSER
				: GLOBAL_DEFS_FOR_TERSER_WITH_AOT),
		};
		return args;
	});

	// todo: re-visit later, disabling by default now
	// config.plugin('DefinePlugin').tap((args) => {
	// 	args[0] = merge(args[0], {
	// 		__UI_USE_EXTERNAL_RENDERER__: true,
	// 	});

	// 	return args;
	// });

	return config;
}

function getAngularCompilerPlugin(): any {
	const { AngularCompilerPlugin } = require('@ngtools/webpack');
	return AngularCompilerPlugin;
}

function getAngularWebpackPlugin(): any {
	const { AngularWebpackPlugin } = require('@ngtools/webpack');
	return AngularWebpackPlugin;
}

function tryRequireResolve(path: string) {
	try {
		return require.resolve(path);
	} catch (err) {
		return null;
	}
}

function getWebpackLoaderPath() {
	return (
		tryRequireResolve(
			'@angular-devkit/build-angular/src/babel/webpack-loader',
		) ??
		tryRequireResolve(
			'@angular-devkit/build-angular/src/tools/babel/webpack-loader',
		) ??
		// fallback to angular 16.1+
		'@angular-devkit/build-angular/src/tools/babel/webpack-loader'
	);
}
