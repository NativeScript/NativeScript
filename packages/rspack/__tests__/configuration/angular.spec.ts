import Config from 'webpack-chain';

import { default as angular } from '../../src/configuration/angular';
import { init } from '../../src';

jest.mock(
	'@ngtools/webpack',
	() => {
		// in tests we mock both plugins so they both show up in the snapshot
		// allowing us to verify the passed configuration is correct.
		class AngularCompilerPlugin {}

		class AngularWebpackPlugin {}

		return {
			AngularCompilerPlugin,
			AngularWebpackPlugin,
		};
	},
	{ virtual: true },
);

describe('angular configuration', () => {
	const platforms = ['ios', 'android'];
	let fsExistsSyncSpy: jest.SpiedFunction<any>;
	let polyfillsPath: string | boolean = false;

	beforeAll(() => {
		const fs = require('fs');
		const original = fs.existsSync;
		fsExistsSyncSpy = jest.spyOn(fs, 'existsSync');

		fsExistsSyncSpy.mockImplementation((path) => {
			if (path === '__jest__/tsconfig.json') {
				return true;
			}

			if (polyfillsPath && path === polyfillsPath) {
				return true;
			}

			return original.call(fs, path);
		});
	});

	afterAll(() => {
		fsExistsSyncSpy.mockRestore();
	});

	for (let platform of platforms) {
		it(`for ${platform}`, () => {
			init({
				[platform]: true,
			});
			expect(angular(new Config()).toString()).toMatchSnapshot();
		});

		it(`loads polyfills.${platform}.ts into the bundle entry if it exists `, () => {
			polyfillsPath = `__jest__/src/polyfills.${platform}.ts`;

			init({
				[platform]: true,
			});
			expect(angular(new Config()).entry('bundle').values()).toMatchSnapshot();

			polyfillsPath = false;
		});
	}

	it(`loads polyfills.ts into the bundle entry if it exists `, () => {
		polyfillsPath = `__jest__/src/polyfills.ts`;

		init({
			ios: true,
		});
		expect(angular(new Config()).entry('bundle').values()).toMatchSnapshot();

		polyfillsPath = false;
	});

	describe('@angular-devkit/build-angular backwards compatible', () => {
		beforeAll(() => {
			jest.mock('@angular-devkit/build-angular/package.json', () => ({
				version: '14.0.0',
			}));
		});

		afterAll(() => {
			jest.unmock('@angular-devkit/build-angular/package.json');
		});

		it('sets scriptTarget for version <15', () => {
			const config = angular(new Config());
			expect(
				config.module
					.rule('angular-webpack-loader')
					.use('webpack-loader')
					.get('options'),
			).toMatchSnapshot();
		});
	});
});
