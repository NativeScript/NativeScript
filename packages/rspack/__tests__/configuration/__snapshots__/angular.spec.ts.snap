// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`angular configuration @angular-devkit/build-angular backwards compatible sets scriptTarget for version <15 1`] = `
{
  "aot": true,
  "scriptTarget": 99,
}
`;

exports[`angular configuration for android 1`] = `
"{
  mode: 'development',
  externals: [
    'package.json',
    '~/package.json'
  ],
  externalsPresets: {
    node: false
  },
  devtool: 'inline-source-map',
  target: 'node',
  watchOptions: {
    ignored: [
      '__jest__/platforms/**',
      '__jest__/App_Resources/**'
    ]
  },
  ignoreWarnings: [
    /System.import\\(\\) is deprecated/,
    /Zone\\.js does not support native async\\/await/,
    /environment(\\.(\\w+))?\\.ts is part of the TypeScript compilation but it's unused/,
    {
      module: /@angular\\/core\\/(__ivy_ngcc__\\/)?fesm2015\\/core.js/,
      message: /Critical dependency: the request of a dependency is an expression/
    },
    /core\\/profiling/,
    /core\\/ui\\/styling/
  ],
  output: {
    path: '__jest__/platforms/android/app/src/main/assets/app',
    pathinfo: false,
    publicPath: '',
    libraryTarget: 'commonjs',
    globalObject: 'global',
    clean: true
  },
  resolve: {
    symlinks: true,
    alias: {
      '~': '__jest__/src',
      '@': '__jest__/src'
    },
    extensions: [
      '.android.ts',
      '.ts',
      '.android.js',
      '.js',
      '.android.mjs',
      '.mjs',
      '.android.css',
      '.css',
      '.android.scss',
      '.scss',
      '.android.json',
      '.json'
    ],
    mainFields: [
      'module',
      'main'
    ],
    modules: [
      '__jest__/node_modules',
      'node_modules'
    ]
  },
  resolveLoader: {
    modules: [
      '__jest__/node_modules/@nativescript/webpack/dist/loaders',
      '__jest__/node_modules',
      'node_modules'
    ]
  },
  module: {
    rules: [
      /* config.module.rule('bundle') */
      {
        enforce: 'post',
        test: '__jest__/src/app.js',
        use: [
          /* config.module.rule('bundle').use('app-css-loader') */
          {
            loader: 'app-css-loader',
            options: {
              platform: 'android'
            }
          }
        ]
      },
      /* config.module.rule('workers') */
      {
        test: /\\.(mjs|js|ts)$/,
        use: [
          /* config.module.rule('workers').use('nativescript-worker-loader') */
          {
            loader: 'nativescript-worker-loader'
          }
        ]
      },
      /* config.module.rule('js') */
      {
        test: /\\.js$/,
        exclude: [
          /node_modules/
        ]
      },
      /* config.module.rule('xml') */
      {
        test: /\\.xml$/,
        use: [
          /* config.module.rule('xml').use('xml-namespace-loader') */
          {
            loader: 'xml-namespace-loader'
          }
        ]
      },
      /* config.module.rule('css') */
      {
        test: /\\.css$/,
        include: [
          '__jest__/src/app.css',
          '__jest__/src/app.android.css',
          /node_modules/
        ],
        use: [
          /* config.module.rule('css').use('apply-css-loader') */
          {
            loader: 'apply-css-loader'
          },
          /* config.module.rule('css').use('css2json-loader') */
          {
            loader: 'css2json-loader'
          },
          /* config.module.rule('css').use('postcss-loader') */
          {
            loader: 'postcss-loader',
            options: {
              postcssOptions: {
                plugins: [
                  [
                    'postcss-import',
                    {
                      resolve: function () { /* omitted long function */ }
                    }
                  ]
                ]
              }
            }
          }
        ]
      },
      /* config.module.rule('scss') */
      {
        test: /\\.scss$/,
        include: [
          '__jest__/src/app.scss',
          '__jest__/src/app.android.scss',
          /node_modules/
        ],
        use: [
          /* config.module.rule('scss').use('apply-css-loader') */
          {
            loader: 'apply-css-loader'
          },
          /* config.module.rule('scss').use('css2json-loader') */
          {
            loader: 'css2json-loader'
          },
          /* config.module.rule('scss').use('postcss-loader') */
          {
            loader: 'postcss-loader',
            options: {
              postcssOptions: {
                plugins: [
                  [
                    'postcss-import',
                    {
                      resolve: function () { /* omitted long function */ }
                    }
                  ]
                ]
              }
            }
          },
          /* config.module.rule('scss').use('sass-loader') */
          {
            loader: 'sass-loader'
          }
        ]
      },
      /* config.module.rule('angular-webpack-loader') */
      {
        test: /\\.[cm]?[tj]sx?$/,
        exclude: [
          /[/\\\\](?:core-js|@babel|tslib|web-animations-js|web-streams-polyfill)[/\\\\]/
        ],
        use: [
          /* config.module.rule('angular-webpack-loader').use('webpack-loader') */
          {
            loader: '<rootDir>/node_modules/@angular-devkit/build-angular/src/tools/babel/webpack-loader.js',
            options: {
              aot: true
            }
          }
        ],
        resolve: {
          fullySpecified: false
        }
      },
      /* config.module.rule('angular') */
      {
        test: /(?:\\.ngfactory.js|\\.ngstyle\\.js|\\.ts)$/,
        use: [
          /* config.module.rule('angular').use('@ngtools/webpack') */
          {
            loader: '@ngtools/webpack'
          }
        ]
      },
      /* config.module.rule('@angular/core') */
      {
        test: /[\\/\\\\]@angular[\\/\\\\]core[\\/\\\\].+\\.js$/,
        parser: {
          system: true
        }
      },
      /* config.module.rule('html') */
      {
        test: /\\.html$/,
        use: [
          /* config.module.rule('html').use('raw-loader') */
          {
            loader: 'raw-loader'
          }
        ]
      },
      /* config.module.rule('css|component') */
      {
        test: /\\.css$/,
        exclude: [
          '__jest__/src/app.css',
          '__jest__/src/app.android.css',
          /node_modules/
        ],
        use: [
          /* config.module.rule('css|component').use('raw-loader') */
          {
            loader: 'raw-loader'
          }
        ]
      },
      /* config.module.rule('scss|component') */
      {
        test: /\\.scss$/,
        exclude: [
          '__jest__/src/app.scss',
          '__jest__/src/app.android.scss',
          /node_modules/
        ],
        use: [
          /* config.module.rule('scss|component').use('raw-loader') */
          {
            loader: 'raw-loader'
          },
          /* config.module.rule('scss|component').use('postcss-loader') */
          {
            loader: 'postcss-loader',
            options: {
              postcssOptions: {
                plugins: [
                  [
                    'postcss-import',
                    {
                      resolve: function () { /* omitted long function */ }
                    }
                  ]
                ]
              }
            }
          },
          /* config.module.rule('scss|component').use('sass-loader') */
          {
            loader: 'sass-loader'
          }
        ]
      }
    ]
  },
  optimization: {
    runtimeChunk: 'single',
    splitChunks: {
      cacheGroups: {
        defaultVendor: {
          test: /[\\\\/]node_modules[\\\\/]/,
          priority: -10,
          name: 'vendor',
          chunks: 'all'
        }
      }
    },
    minimizer: [
      /* config.optimization.minimizer('TerserPlugin') */
      new TerserPlugin(
        {
          terserOptions: {
            compress: {
              collapse_vars: false,
              sequences: false,
              keep_infinity: true,
              drop_console: false,
              global_defs: {
                __UGLIFIED__: true,
                ngDevMode: false,
                ngI18nClosureMode: false,
                ngJitMode: false
              }
            },
            keep_fnames: true,
            keep_classnames: true,
            format: {
              keep_quoted_props: true
            }
          }
        }
      )
    ]
  },
  plugins: [
    /* config.plugin('PlatformSuffixPlugin') */
    new PlatformSuffixPlugin(
      {
        extensions: [
          'android'
        ]
      }
    ),
    /* config.plugin('ContextExclusionPlugin|App_Resources') */
    new ContextExclusionPlugin(
      /(.*)App_Resources(.*)/
    ),
    /* config.plugin('ContextExclusionPlugin|Other_Platforms') */
    new ContextExclusionPlugin(
      /\\.(ios|visionos)\\.(\\w+)$/
    ),
    /* config.plugin('DefinePlugin') */
    new DefinePlugin(
      {
        __DEV__: true,
        __NS_WEBPACK__: true,
        __NS_ENV_VERBOSE__: false,
        __NS_DEV_HOST_IPS__: '["127.0.0.1","192.168.0.10"]',
        __CSS_PARSER__: '"css-tree"',
        __UI_USE_XML_PARSER__: true,
        __UI_USE_EXTERNAL_RENDERER__: false,
        __ANDROID__: true,
        __IOS__: false,
        __VISIONOS__: false,
        __APPLE__: false,
        'global.isAndroid': true,
        'global.isIOS': false,
        'global.isVisionOS': false,
        process: 'global.process'
      }
    ),
    /* config.plugin('CopyWebpackPlugin') */
    new CopyPlugin(
      {
        patterns: [
          {
            from: 'assets/**',
            context: '__jest__/src',
            noErrorOnMissing: true,
            globOptions: {
              dot: false,
              ignore: []
            }
          },
          {
            from: 'fonts/**',
            context: '__jest__/src',
            noErrorOnMissing: true,
            globOptions: {
              dot: false,
              ignore: []
            }
          },
          {
            from: '**/*.+(jpg|png)',
            context: '__jest__/src',
            noErrorOnMissing: true,
            globOptions: {
              dot: false,
              ignore: []
            }
          }
        ]
      }
    ),
    /* config.plugin('WatchStatePlugin') */
    new WatchStatePlugin(),
    /* config.plugin('AngularCompilerPlugin') */
    new AngularCompilerPlugin(
      {
        tsConfigPath: '__jest__/tsconfig.json',
        mainPath: '__jest__/src/app.js',
        forkTypeChecker: false,
        hostReplacementPaths: function () { /* omitted long function */ },
        platformTransformers: [
          function () { /* omitted long function */ }
        ]
      }
    ),
    /* config.plugin('AngularWebpackPlugin') */
    new AngularWebpackPlugin(
      {
        tsconfig: '__jest__/tsconfig.json',
        directTemplateLoading: false,
        jitMode: false
      }
    )
  ],
  entry: {
    bundle: [
      '@nativescript/core/globals/index',
      '@nativescript/core/bundle-entry-points',
      '__jest__/src/app.js',
      '@nativescript/core/ui/frame',
      '@nativescript/core/ui/frame/activity'
    ]
  }
}"
`;

exports[`angular configuration for ios 1`] = `
"{
  mode: 'development',
  externals: [
    'package.json',
    '~/package.json'
  ],
  externalsPresets: {
    node: false
  },
  devtool: 'inline-source-map',
  target: 'node',
  watchOptions: {
    ignored: [
      '__jest__/platforms/**',
      '__jest__/App_Resources/**'
    ]
  },
  ignoreWarnings: [
    /System.import\\(\\) is deprecated/,
    /Zone\\.js does not support native async\\/await/,
    /environment(\\.(\\w+))?\\.ts is part of the TypeScript compilation but it's unused/,
    {
      module: /@angular\\/core\\/(__ivy_ngcc__\\/)?fesm2015\\/core.js/,
      message: /Critical dependency: the request of a dependency is an expression/
    },
    /core\\/profiling/,
    /core\\/ui\\/styling/
  ],
  output: {
    path: '__jest__/platforms/ios/jest/app',
    pathinfo: false,
    publicPath: '',
    libraryTarget: 'commonjs',
    globalObject: 'global',
    clean: true
  },
  resolve: {
    symlinks: true,
    alias: {
      '~': '__jest__/src',
      '@': '__jest__/src'
    },
    extensions: [
      '.ios.ts',
      '.ts',
      '.ios.js',
      '.js',
      '.ios.mjs',
      '.mjs',
      '.ios.css',
      '.css',
      '.ios.scss',
      '.scss',
      '.ios.json',
      '.json'
    ],
    mainFields: [
      'module',
      'main'
    ],
    modules: [
      '__jest__/node_modules',
      'node_modules'
    ]
  },
  resolveLoader: {
    modules: [
      '__jest__/node_modules/@nativescript/webpack/dist/loaders',
      '__jest__/node_modules',
      'node_modules'
    ]
  },
  module: {
    rules: [
      /* config.module.rule('bundle') */
      {
        enforce: 'post',
        test: '__jest__/src/app.js',
        use: [
          /* config.module.rule('bundle').use('app-css-loader') */
          {
            loader: 'app-css-loader',
            options: {
              platform: 'ios'
            }
          }
        ]
      },
      /* config.module.rule('workers') */
      {
        test: /\\.(mjs|js|ts)$/,
        use: [
          /* config.module.rule('workers').use('nativescript-worker-loader') */
          {
            loader: 'nativescript-worker-loader'
          }
        ]
      },
      /* config.module.rule('js') */
      {
        test: /\\.js$/,
        exclude: [
          /node_modules/
        ]
      },
      /* config.module.rule('xml') */
      {
        test: /\\.xml$/,
        use: [
          /* config.module.rule('xml').use('xml-namespace-loader') */
          {
            loader: 'xml-namespace-loader'
          }
        ]
      },
      /* config.module.rule('css') */
      {
        test: /\\.css$/,
        include: [
          '__jest__/src/app.css',
          '__jest__/src/app.ios.css',
          /node_modules/
        ],
        use: [
          /* config.module.rule('css').use('apply-css-loader') */
          {
            loader: 'apply-css-loader'
          },
          /* config.module.rule('css').use('css2json-loader') */
          {
            loader: 'css2json-loader'
          },
          /* config.module.rule('css').use('postcss-loader') */
          {
            loader: 'postcss-loader',
            options: {
              postcssOptions: {
                plugins: [
                  [
                    'postcss-import',
                    {
                      resolve: function () { /* omitted long function */ }
                    }
                  ]
                ]
              }
            }
          }
        ]
      },
      /* config.module.rule('scss') */
      {
        test: /\\.scss$/,
        include: [
          '__jest__/src/app.scss',
          '__jest__/src/app.ios.scss',
          /node_modules/
        ],
        use: [
          /* config.module.rule('scss').use('apply-css-loader') */
          {
            loader: 'apply-css-loader'
          },
          /* config.module.rule('scss').use('css2json-loader') */
          {
            loader: 'css2json-loader'
          },
          /* config.module.rule('scss').use('postcss-loader') */
          {
            loader: 'postcss-loader',
            options: {
              postcssOptions: {
                plugins: [
                  [
                    'postcss-import',
                    {
                      resolve: function () { /* omitted long function */ }
                    }
                  ]
                ]
              }
            }
          },
          /* config.module.rule('scss').use('sass-loader') */
          {
            loader: 'sass-loader'
          }
        ]
      },
      /* config.module.rule('angular-webpack-loader') */
      {
        test: /\\.[cm]?[tj]sx?$/,
        exclude: [
          /[/\\\\](?:core-js|@babel|tslib|web-animations-js|web-streams-polyfill)[/\\\\]/
        ],
        use: [
          /* config.module.rule('angular-webpack-loader').use('webpack-loader') */
          {
            loader: '<rootDir>/node_modules/@angular-devkit/build-angular/src/tools/babel/webpack-loader.js',
            options: {
              aot: true
            }
          }
        ],
        resolve: {
          fullySpecified: false
        }
      },
      /* config.module.rule('angular') */
      {
        test: /(?:\\.ngfactory.js|\\.ngstyle\\.js|\\.ts)$/,
        use: [
          /* config.module.rule('angular').use('@ngtools/webpack') */
          {
            loader: '@ngtools/webpack'
          }
        ]
      },
      /* config.module.rule('@angular/core') */
      {
        test: /[\\/\\\\]@angular[\\/\\\\]core[\\/\\\\].+\\.js$/,
        parser: {
          system: true
        }
      },
      /* config.module.rule('html') */
      {
        test: /\\.html$/,
        use: [
          /* config.module.rule('html').use('raw-loader') */
          {
            loader: 'raw-loader'
          }
        ]
      },
      /* config.module.rule('css|component') */
      {
        test: /\\.css$/,
        exclude: [
          '__jest__/src/app.css',
          '__jest__/src/app.ios.css',
          /node_modules/
        ],
        use: [
          /* config.module.rule('css|component').use('raw-loader') */
          {
            loader: 'raw-loader'
          }
        ]
      },
      /* config.module.rule('scss|component') */
      {
        test: /\\.scss$/,
        exclude: [
          '__jest__/src/app.scss',
          '__jest__/src/app.ios.scss',
          /node_modules/
        ],
        use: [
          /* config.module.rule('scss|component').use('raw-loader') */
          {
            loader: 'raw-loader'
          },
          /* config.module.rule('scss|component').use('postcss-loader') */
          {
            loader: 'postcss-loader',
            options: {
              postcssOptions: {
                plugins: [
                  [
                    'postcss-import',
                    {
                      resolve: function () { /* omitted long function */ }
                    }
                  ]
                ]
              }
            }
          },
          /* config.module.rule('scss|component').use('sass-loader') */
          {
            loader: 'sass-loader'
          }
        ]
      }
    ]
  },
  optimization: {
    runtimeChunk: 'single',
    splitChunks: {
      cacheGroups: {
        defaultVendor: {
          test: /[\\\\/]node_modules[\\\\/]/,
          priority: -10,
          name: 'vendor',
          chunks: 'all'
        }
      }
    },
    minimizer: [
      /* config.optimization.minimizer('TerserPlugin') */
      new TerserPlugin(
        {
          terserOptions: {
            compress: {
              collapse_vars: true,
              sequences: true,
              keep_infinity: true,
              drop_console: false,
              global_defs: {
                __UGLIFIED__: true,
                ngDevMode: false,
                ngI18nClosureMode: false,
                ngJitMode: false
              }
            },
            keep_fnames: true,
            keep_classnames: true,
            format: {
              keep_quoted_props: true
            }
          }
        }
      )
    ]
  },
  plugins: [
    /* config.plugin('PlatformSuffixPlugin') */
    new PlatformSuffixPlugin(
      {
        extensions: [
          'ios'
        ]
      }
    ),
    /* config.plugin('ContextExclusionPlugin|App_Resources') */
    new ContextExclusionPlugin(
      /(.*)App_Resources(.*)/
    ),
    /* config.plugin('ContextExclusionPlugin|Other_Platforms') */
    new ContextExclusionPlugin(
      /\\.(android|visionos)\\.(\\w+)$/
    ),
    /* config.plugin('DefinePlugin') */
    new DefinePlugin(
      {
        __DEV__: true,
        __NS_WEBPACK__: true,
        __NS_ENV_VERBOSE__: false,
        __NS_DEV_HOST_IPS__: '["127.0.0.1","192.168.0.10"]',
        __CSS_PARSER__: '"css-tree"',
        __UI_USE_XML_PARSER__: true,
        __UI_USE_EXTERNAL_RENDERER__: false,
        __ANDROID__: false,
        __IOS__: true,
        __VISIONOS__: false,
        __APPLE__: true,
        'global.isAndroid': false,
        'global.isIOS': true,
        'global.isVisionOS': false,
        process: 'global.process'
      }
    ),
    /* config.plugin('CopyWebpackPlugin') */
    new CopyPlugin(
      {
        patterns: [
          {
            from: 'assets/**',
            context: '__jest__/src',
            noErrorOnMissing: true,
            globOptions: {
              dot: false,
              ignore: []
            }
          },
          {
            from: 'fonts/**',
            context: '__jest__/src',
            noErrorOnMissing: true,
            globOptions: {
              dot: false,
              ignore: []
            }
          },
          {
            from: '**/*.+(jpg|png)',
            context: '__jest__/src',
            noErrorOnMissing: true,
            globOptions: {
              dot: false,
              ignore: []
            }
          }
        ]
      }
    ),
    /* config.plugin('WatchStatePlugin') */
    new WatchStatePlugin(),
    /* config.plugin('AngularCompilerPlugin') */
    new AngularCompilerPlugin(
      {
        tsConfigPath: '__jest__/tsconfig.json',
        mainPath: '__jest__/src/app.js',
        forkTypeChecker: false,
        hostReplacementPaths: function () { /* omitted long function */ },
        platformTransformers: [
          function () { /* omitted long function */ }
        ]
      }
    ),
    /* config.plugin('AngularWebpackPlugin') */
    new AngularWebpackPlugin(
      {
        tsconfig: '__jest__/tsconfig.json',
        directTemplateLoading: false,
        jitMode: false
      }
    )
  ],
  entry: {
    bundle: [
      '@nativescript/core/globals/index',
      '@nativescript/core/bundle-entry-points',
      '__jest__/src/app.js'
    ],
    'tns_modules/inspector_modules': [
      '@nativescript/core/inspector_modules'
    ]
  }
}"
`;

exports[`angular configuration loads polyfills.android.ts into the bundle entry if it exists  1`] = `
[
  "__jest__/src/polyfills.android.ts",
  "@nativescript/core/bundle-entry-points",
  "__jest__/src/app.js",
  "@nativescript/core/ui/frame",
  "@nativescript/core/ui/frame/activity",
]
`;

exports[`angular configuration loads polyfills.ios.ts into the bundle entry if it exists  1`] = `
[
  "__jest__/src/polyfills.ios.ts",
  "@nativescript/core/bundle-entry-points",
  "__jest__/src/app.js",
]
`;

exports[`angular configuration loads polyfills.ts into the bundle entry if it exists  1`] = `
[
  "__jest__/src/polyfills.ts",
  "@nativescript/core/bundle-entry-points",
  "__jest__/src/app.js",
]
`;
